### PR TITLE
Fix relationship between node service and mounter interface

### PR DESCRIPTION
**Pull request by torredil**
**What is this PR about? / Why do we need it?**

This patch is 2/2, designed to improve the architecture of the code base with the goal of promoting a more cohesive and understandable design that makes it easier to reason about the driver and make future modifications or extensions more intuitive.

More specifically, this PR fixes the relationship between the node service and mounter interface, allowing for a platform agnostic node service.

**Summary of Changes**

- Integrate `resizeFs` functionality into mounter package
   - Updated the `NodeStageVolume` and `NodeExpandVolume` RPCs to use the `Resize` method directly from the `Mounter` interface, instead of creating a new `ResizeFs` instance.
 - Remove `DeviceIdentifier` interface
   - Replaced the usage of` deviceIdentifier` with direct calls to `os.Lstat` and `filepath.EvalSymlinks`.
 - Refactor node service to be platform agnostic
   - Moved platform-specific code from `node_linux.go` and `node_windows.go` to `mount_linux.go` and `mount_windows.go`.
 - Refactor node service initialization and dependency injection
   - Improved modularity and testability by injecting dependencies (`metadata.MetadataService`, `Mounter`, `metadata.KubernetesAPIClient`) into the `NewDriver()` and `NewNodeService()` functions.
   

**What testing is done?**

- manual
- `make test && make verify`
- CI

### CODE REVIEW

File: pkg/driver/mount.go [hunk 1]
```diff
@@ -51,7 +51,7 @@ type NodeMounter struct {
 	*mountutils.SafeFormatAndMount
 }
 
-func newNodeMounter() (Mounter, error) {
+func NewNodeMounter() (Mounter, error) {
```
[the remark this answers is not in the document]
Comment (torredil) [reply pkg/driver/mount.go:53]: Added comment ✅

File: pkg/driver/mount_windows.go [hunk 2]
```diff
@@ -20,13 +20,94 @@ limitations under the License.
...
+	"strings"
 
+	diskapi "github.com/kubernetes-csi/csi-proxy/client/api/disk/v1"
+	diskclient "github.com/kubernetes-csi/csi-proxy/client/groups/disk/v1"
 	"github.com/kubernetes-sigs/aws-ebs-csi-driver/pkg/mounter"
+	"k8s.io/klog/v2"
 	mountutils "k8s.io/mount-utils"
 )
 
+func (m NodeMounter) FindDevicePath(devicePath, volumeID, _, _ string) (string, error) {
```
[the remark this answers is not in the document]
Comment (torredil) [reply pkg/driver/mount_windows.go:39]: Added comments ✅

File: pkg/driver/mount_linux.go [hunk 3]
```diff
@@ -22,12 +22,187 @@ package driver
...
 	"strconv"
 	"strings"
 
+	"github.com/kubernetes-sigs/aws-ebs-csi-driver/pkg/util"
+	"golang.org/x/sys/unix"
+	"k8s.io/klog/v2"
 	mountutils "k8s.io/mount-utils"
 )
 
+func (m *NodeMounter) FindDevicePath(devicePath, volumeID, partition, region string) (string, error) {
```
[the remark this answers is not in the document]
Comment (torredil) [reply pkg/driver/mount_linux.go]: Added comments ✅

File: pkg/driver/mount.go [hunk 4]
```diff
@@ -41,11 +38,11 @@ type Mounter interface {
...
 	Unpublish(path string) error
 	Unstage(path string) error
-	NewResizeFs() (Resizefs, error)
-}
-
-type Resizefs interface {
 	Resize(devicePath, deviceMountPath string) (bool, error)
+	FindDevicePath(devicePath, volumeID, partition, region string) (string, error)
+	PreparePublishTarget(target string) error
+	IsBlockDevice(fullPath string) (bool, error)
```
[the remark this answers is not in the document]
Comment (torredil) [reply pkg/driver/mount.go:43]: > Are we sure we should be sticking everything into one interface? (It does seem that the convenience of not worrying about Linux vs Windows at call-site in Node and in unit tests is worth it. And a single set-and-forget singleton makes this easy)

Yes exactly, that would be my recommendation.

> If we are sure, perhaps we should consider changing the interface name from Mounter to DeviceUtils or OSSpecificUtils or something that encapsulates the sheer diversity of functions?

I don't have a strong preference here, open to this recommendation for renaming the interface cc @ConnorJC3 thoughts?

File: pkg/driver/driver.go [hunk 5]
```diff
@@ -45,102 +45,52 @@ const (
...
-		driver.controllerService = newControllerService(&driverOptions)
+		driver.controller = NewControllerService(c, o)
 	case NodeMode:
-		driver.nodeService = newNodeService(&driverOptions)
+		driver.node = NewNodeService(o, md, m, k)
 	case AllMode:
-		driver.controllerService = newControllerService(&driverOptions)
-		driver.nodeService = newNodeService(&driverOptions)
+		driver.controller = NewControllerService(c, o)
+		driver.node = NewNodeService(o, md, m, k)
```
[the remark this answers is not in the document]
Comment (torredil) [reply pkg/driver/driver.go:85]: 🍲

File: pkg/mounter/mount_linux.go [hunk 6]
```diff
@@ -82,55 +70,76 @@ func (d *nodeService) findDevicePath(devicePath, volumeID, partition string) (st
...
 		klog.V(5).InfoS("[Debug] error searching for nvme path", "nvmeName", nvmeName, "err", err)
 	}
 
-	if util.IsSBE(d.metadata.GetRegion()) {
+	if util.IsSBE(region) {
 		klog.V(5).InfoS("[Debug] Falling back to snow volume lookup", "devicePath", devicePath)
-		// Snow completely ignores the requested device path and mounts volumes starting at /dev/vda .. /dev/vdb .. etc
-		// Morph the device path to the snow form by chopping off the last letter and prefixing with /dev/vd
-		// VMs on snow devices are currently limited to 10 block devices each - if that ever exceeds 26 this will need
-		// to be adapted
```
Comment (ConnorJC3) [inline pkg/mounter/mount_linux.go:116]: Why remove this comment?

File: hack/update-mockgen.sh [hunk 7]
```diff
@@ -21,17 +21,12 @@ BIN="$(dirname "$(realpath "${BASH_SOURCE[0]}")")/../bin"
 # Source-based mocking for internal interfaces
 "${BIN}/mockgen" -package cloud -destination=./pkg/cloud/mock_cloud.go -source pkg/cloud/interface.go
 "${BIN}/mockgen" -package metadata -destination=./pkg/cloud/metadata/mock_metadata.go -source pkg/cloud/metadata/interface.go
```
Comment (ConnorJC3) [inline hack/update-mockgen.sh:23]: Is there a reason we can't use a `mocks` folder for cloud/metadata mocks? I think we should be consistent and do it everywhere or nowhere

File: cmd/main.go [hunk 8]
```diff
@@ -132,19 +133,22 @@ func main() {
...
+		K8sAPIClient:      metadata.DefaultKubernetesAPIClient,
+	}
+
 	region := os.Getenv("AWS_REGION")
+
+	md, metadataErr := metadata.NewMetadataService(cfg, region)
+	if metadataErr != nil {
+		klog.ErrorS(err, "Could not determine region from any metadata service. The region can be manually supplied via the AWS_REGION environment variable.")
+		panic(err)
+	}
```
Comment (ConnorJC3) [inline cmd/main.go:147]: Why is this moved? Moving it out of the `if region == ""` statement breaks the ability to set the region via `AWS_REGION` when metadata is not available.

File: Makefile [hunk 9]
```diff
@@ -74,7 +74,7 @@ clean:
 
 .PHONY: test
 test:
-	go test -v -race ./cmd/... ./pkg/...
+	go test -race ./cmd/... ./pkg/...
```
Comment (ConnorJC3) [inline Makefile:77]: Drop this diff

Comment (torredil) [reply Makefile:77]: Fixed

File: cmd/main.go (hunk 8 above)
Comment (torredil) [reply cmd/main.go:147]: The node service heavily relies on metadata.

As an example, see NodeGetInfo:
- `zone := d.metadata.GetAvailabilityZone()`
- `outpostArn := d.metadata.GetOutpostArn()`
-  `d.metadata.GetInstanceID()`

It was moved out of that conditional so that it may be passed into `NewDriver`.

File: hack/update-mockgen.sh (hunk 7 above)
Comment (torredil) [reply hack/update-mockgen.sh:23]: Tried this but was unable to do so due to a circular dependency. I think having a `mocks` directory for the driver package is an organizational improvement but inclined to remove it if you feel strongly about the request.

File: pkg/mounter/mount_linux.go (hunk 6 above)
Comment (torredil) [reply pkg/mounter/mount_linux.go:116]: Added comment

File: cmd/main.go (hunk 8 above)
Comment (ConnorJC3) [reply cmd/main.go:147]: The code needs to be reworked so you can pass `AWS_REGION` when running the controller then, as that is supported today and we likely have at least a few users using it

File: hack/update-mockgen.sh (hunk 7 above)
Comment (ConnorJC3) [reply hack/update-mockgen.sh:23]: I feel strongly that we should be consistent, I don't really care which way it is.

File: pkg/driver/mount.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (ConnorJC3) [reply pkg/driver/mount.go:43]: Talked about this offline - we decided to keep the `mounter` name for now as (1) we don't have an obviously better option and (2) it accurately describes the majority of this package's usecase (mounting and interacting with mounts).

File: cmd/main.go (hunk 8 above)
Comment (torredil) [reply cmd/main.go]: Updated

File: hack/update-mockgen.sh (hunk 7 above)
Comment (torredil) [reply hack/update-mockgen.sh:23]: Removed mocks directory

File: pkg/driver/node.go [hunk 10]
```diff
@@ -81,53 +82,35 @@ var (
...
-	nodeMounter, err := newNodeMounter()
-	if err != nil {
-		panic(err)
-	}
-
 	// Remove taint from node to indicate driver startup success
 	// This is done at the last possible moment to prevent race conditions or false positive removals
 	time.AfterFunc(taintRemovalInitialDelay, func() {
-		removeTaintInBackground(cfg.K8sAPIClient, removeNotReadyTaint)
+		removeTaintInBackground(k, taintRemovalBackoff, removeNotReadyTaint)
```
[the remark this answers is not in the document]
Comment (torredil) [reply pkg/driver/node.go:102]: Already done:

```
	t.Run("Retries exhausted", func(t *testing.T) {
		mockRemovalCount := 0
		mockRemovalFunc := func(_ kubernetes.Interface) error {
			mockRemovalCount += 1
			return fmt.Errorf("Taint removal failed!")
		}
		removeTaintInBackground(nil, wait.Backoff{
			Steps:    5,
			Duration: 1 * time.Millisecond,
		}, mockRemovalFunc)
		assert.Equal(t, 5, mockRemovalCount)
	})
}
```

File: Makefile [hunk 11]
```diff
@@ -74,7 +74,7 @@ clean:
 
 .PHONY: test
 test:
-	go test -v -race ./cmd/... ./pkg/...
```
[the remark this answers is not in the document]
Comment (torredil) [reply Makefile]: Addressed in a previous comment

File: cmd/main.go [hunk 12]
```diff
@@ -153,7 +157,17 @@ func main() {
...
 
-	drv, err := driver.NewDriver(cloud, &options)
+	m, err := mounter.NewNodeMounter()
+	if err != nil {
+		panic(err)
+	}
+
+	k8sClient, err := cfg.K8sAPIClient()
+	if err != nil {
+		klog.V(4).InfoS("Failed to setup k8s client")
```
[the remark this answers is not in the document]
Comment (torredil) [reply cmd/main.go:178]: Updated to v = 2

File: pkg/mounter/mount_linux.go [hunk 13]
```diff
@@ -31,46 +31,34 @@ import (
...
-// findDevicePath finds path of device and verifies its existence
+// FindDevicePath finds path of device and verifies its existence
 // if the device is not nvme, return the path directly
 // if the device is nvme, finds and returns the nvme device path eg. /dev/nvme1n1
-func (d *nodeService) findDevicePath(devicePath, volumeID, partition string) (string, error) {
+func (m *NodeMounter) FindDevicePath(devicePath, volumeID, partition, region string) (string, error) {
 	strippedVolumeName := strings.Replace(volumeID, "-", "", -1)
 	canonicalDevicePath := ""
 
-	// If the given path exists, the device MAY be nvme. Further, it MAY be a
```
[the remark this answers is not in the document]
Comment (torredil) [reply pkg/mounter/mount_linux.go]: Added comments

**Review by ConnorJC3 — APPROVED** [review]
/lgtm

**Review by AndrewSirenko — APPROVED** [review]
